lib/optimize/CommonsChunkPlugin.js

```diff
@@ -2,183 +2,176 @@
+							asyncOption || chunk.parents.length === 1);
+					} else {
+						if(commonChunk.parents.length > 0) {
+							compilation.errors.push(new Error("CommonsChunkPlugin: While running in normal mode it's not allowed to use a non-entry chunk (" + commonChunk.name + ")"));
 							return;
 						}
-						if(module.chunkCondition && !module.chunkCondition(commonChunk))
-							return;
-						reallyUsedModules.push(module);
-					});
-				}
-				if(minSize) {
-					var size = reallyUsedModules.reduce(function(a, b) {
-						return a + b.size();
-					}, 0);
-					if(size < minSize)
-						return;
-				}
-				var reallyUsedChunks = [];
-				reallyUsedModules.forEach(function(module) {
-					usedChunks.forEach(function(chunk) {
-						if(module.removeChunk(chunk)) {
-							if(reallyUsedChunks.indexOf(chunk) < 0)
-								reallyUsedChunks.push(chunk);
-						}
-					});
-					commonChunk.addModule(module);
-					module.addChunk(commonChunk);
-				});
-				if(asyncOption) {
-					reallyUsedChunks.forEach(function(chunk) {
-						if(chunk.isInitial())
+						usedChunks = chunks.filter(chunk => commonChunks.indexOf(chunk) < idx && chunk.hasRuntime())
+					}
+					let asyncChunk;
+					if(asyncOption) {
+						asyncChunk = compilation.addChunk(typeof asyncOption === "string" ? asyncOption : undefined);
+						asyncChunk.chunkReason = "async commons chunk";
+						asyncChunk.extraAsync = true;
+						asyncChunk.addParent(commonChunk);
+						commonChunk.addChunk(asyncChunk);
+						commonChunk = asyncChunk;
+					}
+					const reallyUsedModules = [];
+					if(minChunks !== Infinity) {
+						const commonModulesCount = [];
+						const commonModules = [];
+						usedChunks.forEach((chunk) => {
+							chunk.modules.forEach((module) => {
+								const idx = commonModules.indexOf(module);
+								if(idx < 0) {
+									commonModules.push(module);
+									commonModulesCount.push(1);
+								} else {
+									commonModulesCount[idx]++;
+								}
+							});
+						});
+						const _minChunks = (minChunks || Math.max(2, usedChunks.length))
+						commonModulesCount.forEach((count, idx) => {
+							const module = commonModules[idx];
+							if(typeof minChunks === "function") {
+								if(!minChunks(module, count))
+									return;
+							} else if(count < _minChunks) {
+								return;
+							}
+							if(module.chunkCondition && !module.chunkCondition(commonChunk))
+								return;
+							reallyUsedModules.push(module);
+						});
+					}
+					if(minSize) {
+						const size = reallyUsedModules.reduce((a, b) => a + b.size(), 0);
+						if(size < minSize)
 							return;
-						chunk.blocks.forEach(function(block) {
-							block.chunks.unshift(commonChunk);
-							commonChunk.addBlock(block);
+					}
+					const reallyUsedChunks = [];
+					reallyUsedModules.forEach((module) => {
+						usedChunks.forEach((chunk) => {
+							if(module.removeChunk(chunk)) {
+								if(reallyUsedChunks.indexOf(chunk) < 0)
```

Comment:
Looks like `reallyUsedChunks` should be a `Set`

```js
const reallyUsedChunks = new Set()
reallyUsedModules.forEach(module => {
  usedChunks.forEach(chunk => {
    if(module.removeChunk(chunk)) {
      reallyUsedChunks.add(chunk);
    }
  });
  /* ... */
});
if(asyncOption) {
  for(const chunk of reallyUsedChunks) {
    if (chunk.isInitial()) continue;
    /* ... */
  }
  const asyncChunk.origins = Array.from(reallyUsedChunks).map(/*... */)/* ... */
}
/* ... */
```

The result of the `Array.from` could also just be saved to a `const` inside the `if`, to keep using it as array both times.

WTB iterator map/filter/reduce...